File: maplibre_gl/ios/maplibre_gl/Sources/maplibre_gl/EditablePolylineRenderer.swift
import Foundation
import MapLibre
import UIKit

/**
 * Manages visual feedback for polyline editing operations using MLNAnnotationView customization.
 * 
 * This class handles:
 * - Break point marker styling and rendering
 * - Preview line rendering during drag operations with MLNPolyline styling
 * - Visual state management for active editing sessions
 * - Smooth Core Animation transitions for editing operations
 */
class EditablePolylineRenderer {
    private static let TAG = "EditablePolylineRenderer"
    
    private let mapView: MLNMapView
    private var breakPointAnnotations: [String: MLNPointAnnotation] = [:]
    private var previewLines: [String: MLNPolyline] = [:]
    private var editingStyle: [String: Any] = [:]
    
    // Source and layer IDs for rendering
    private let breakPointSourceId = "polyline-editing-breakpoints"
    private let previewLineSourceId = "polyline-editing-preview"
    private let breakPointLayerId = "polyline-editing-breakpoint-layer"
    private let previewLineLayerId = "polyline-editing-preview-layer"
    
    init(mapView: MLNMapView) {
        self.mapView = mapView
        setupDefaultStyle()
        NSLog("\(EditablePolylineRenderer.TAG): Initialized editable polyline renderer")
    }
    
    /**
     * Sets up default visual styling for editing elements.
     */
    private func setupDefaultStyle() {
        editingStyle = [
            "breakPointColor": "#FF0000",
            "breakPointRadius": 8.0,
            "breakPointBorderColor": "#FFFFFF",
            "breakPointBorderWidth": 2.0,
            "previewLineColor": "#00FF00",
            "previewLineOpacity": 0.7,
            "previewLineWidth": 3.0
        ]
    }
    
    /**
     * Initializes the renderer after the map style is loaded.
     * This should be called when the map style is ready.
     */
    func initialize() {
        setupBreakPointLayer()
        setupPreviewLineLayer()
        NSLog("\(EditablePolylineRenderer.TAG): Renderer initialized with map style")
    }
    
    /**
     * Sets up the layer for rendering break point markers.
     */
    private func setupBreakPointLayer() {
        guard let style = mapView.style else {
            NSLog("\(EditablePolylineRenderer.TAG): Cannot setup break point layer - map style not available")
            return
        }
        
        // Create source for break points
        let breakPointSource = MLNShapeSource(identifier: breakPointSourceId, shape: nil, options: nil)
        style.addSource(breakPointSource)
        
        // Create layer for break points
        let breakPointLayer = MLNCircleStyleLayer(identifier: breakPointLayerId, source: breakPointSource)
        
        // Apply styling
        let radius = editingStyle["breakPointRadius"] as? Double ?? 8.0
        let color = parseColor(editingStyle["breakPointColor"] as? String ?? "#FF0000")
        let borderColor = parseColor(editingStyle["breakPointBorderColor"] as? String ?? "#FFFFFF")
        let borderWidth = editingStyle["breakPointBorderWidth"] as? Double ?? 2.0
        
        breakPointLayer.circleRadius = NSExpression(forConstantValue: radius)
        breakPointLayer.circleColor = NSExpression(forConstantValue: color)
        breakPointLayer.circleStrokeColor = NSExpression(forConstantValue: borderColor)
        breakPointLayer.circleStrokeWidth = NSExpression(forConstantValue: borderWidth)
        
        style.addLayer(breakPointLayer)
        NSLog("\(EditablePolylineRenderer.TAG): Break point layer setup complete")
    }
    
    /**
     * Sets up the layer for rendering preview lines.
     */
    private func setupPreviewLineLayer() {
        guard let style = mapView.style else {
            NSLog("\(EditablePolylineRenderer.TAG): Cannot setup preview line layer - map style not available")
            return
        }
        
        // Create source for preview lines
        let previewLineSource = MLNShapeSource(identifier: previewLineSourceId, shape: nil, options: nil)
        style.addSource(previewLineSource)
        
        // Create layer for preview lines
        let previewLineLayer = MLNLineStyleLayer(identifier: previewLineLayerId, source: previewLineSource)
        
        // Apply styling
        let color = parseColor(editingStyle["previewLineColor"] as? String ?? "#00FF00")
        let opacity = editingStyle["previewLineOpacity"] as? Double ?? 0.7
        let width = editingStyle["previewLineWidth"] as? Double ?? 3.0
        
        previewLineLayer.lineColor = NSExpression(forConstantValue: color)
        previewLineLayer.lineOpacity = NSExpression(forConstantValue: opacity)
        previewLineLayer.lineWidth = NSExpression(forConstantValue: width)
        previewLineLayer.lineCap = NSExpression(forConstantValue: "round")
        previewLineLayer.lineJoin = NSExpression(forConstantValue: "round")
        
        style.addLayer(previewLineLayer)
        NSLog("\(EditablePolylineRenderer.TAG): Preview line layer setup complete")
    }
    
    /**
     * Shows a break point marker at the specified coordinate.
     *
     * @param lineId The ID of the line the break point belongs to
     * @param coordinate The coordinate where the break point should be displayed
     */
    func showBreakPoint(lineId: String, coordinate: CLLocationCoordinate2D) {
        // Create point annotation for the break point
        let annotation = MLNPointAnnotation()
        annotation.coordinate = coordinate
        annotation.title = "Break Point"
        
        // Store the annotation
        breakPointAnnotations[lineId] = annotation
        
        // Update the break point source
        updateBreakPointSource()
        
        NSLog("\(EditablePolylineRenderer.TAG): Showing break point for line \(lineId) at \(coordinate.latitude), \(coordinate.longitude)")
    }
    
    /**
     * Hides the break point marker for the specified line.
     *
     * @param lineId The ID of the line to hide the break point for
     */
    func hideBreakPoint(lineId: String) {
        breakPointAnnotations.removeValue(forKey: lineId)
        updateBreakPointSource()
        NSLog("\(EditablePolylineRenderer.TAG): Hidden break point for line \(lineId)")
    }
    
    /**
     * Shows a preview line during drag operations.
     *
     * @param lineId The ID of the line being edited
     * @param coordinates The coordinates of the preview line
     */
    func showPreviewLine(lineId: String, coordinates: [CLLocationCoordinate2D]) {
        guard coordinates.count >= 2 else {
            NSLog("\(EditablePolylineRenderer.TAG): Cannot show preview line with less than 2 coordinates")
            return
        }
        
        // Create polyline for preview
        let polyline = MLNPolyline(coordinates: coordinates, count: UInt(coordinates.count))
        previewLines[lineId] = polyline
        
        // Update the preview line source
        updatePreviewLineSource()
        
        NSLog("\(EditablePolylineRenderer.TAG): Showing preview line for line \(lineId) with \(coordinates.count) coordinates")
    }
    
    /**
     * Hides the preview line for the specified line.
     *
     * @param lineId The ID of the line to hide the preview line for
     */
    func hidePreviewLine(lineId: String) {
        previewLines.removeValue(forKey: lineId)
        updatePreviewLineSource()
        NSLog("\(EditablePolylineRenderer.TAG): Hidden preview line for line \(lineId)")
    }
    
    /**
     * Updates the visual styling for editing elements.
     *
     * @param style A dictionary containing the new style properties
     */
    func updateStyle(_ style: [String: Any]) {
        NSLog("\(EditablePolylineRenderer.TAG): Updating style with \(style.count) properties")
        
        // Merge with existing style
        for (key, value) in style {
            editingStyle[key] = value
        }
        
        // Update layer styles
        updateBreakPointLayerStyle()
        updatePreviewLineLayerStyle()
        
        NSLog("\(EditablePolylineRenderer.TAG): Style update complete")
    }
    
    /**
     * Updates the break point layer styling.
     */
    private func updateBreakPointLayerStyle() {
        guard let style = mapView.style,
              let layer = style.layer(withIdentifier: breakPointLayerId) as? MLNCircleStyleLayer else {
            return
        }
        
        let radius = editingStyle["breakPointRadius"] as? Double ?? 8.0
        let color = parseColor(editingStyle["breakPointColor"] as? String ?? "#FF0000")
        let borderColor = parseColor(editingStyle["breakPointBorderColor"] as? String ?? "#FFFFFF")
        let borderWidth = editingStyle["breakPointBorderWidth"] as? Double ?? 2.0
        
        layer.circleRadius = NSExpression(forConstantValue: radius)
        layer.circleColor = NSExpression(forConstantValue: color)
        layer.circleStrokeColor = NSExpression(forConstantValue: borderColor)
        layer.circleStrokeWidth = NSExpression(forConstantValue: borderWidth)
    }
    
    /**
     * Updates the preview line layer styling.
     */
    private func updatePreviewLineLayerStyle() {
        guard let style = mapView.style,
              let layer = style.layer(withIdentifier: previewLineLayerId) as? MLNLineStyleLayer else {
            return
        }
        
        let color = parseColor(editingStyle["previewLineColor"] as? String ?? "#00FF00")
        let opacity = editingStyle["previewLineOpacity"] as? Double ?? 0.7
        let width = editingStyle["previewLineWidth"] as? Double ?? 3.0
        
        layer.lineColor = NSExpression(forConstantValue: color)
        layer.lineOpacity = NSExpression(forConstantValue: opacity)
        layer.lineWidth = NSExpression(forConstantValue: width)
    }
    
    /**
     * Updates the break point source with current annotations.
     */
    private func updateBreakPointSource() {
        guard let style = mapView.style,
              let source = style.source(withIdentifier: breakPointSourceId) as? MLNShapeSource else {
            return
        }
        
        if breakPointAnnotations.isEmpty {
            source.shape = nil
        } else {
            let annotations = Array(breakPointAnnotations.values)
            let shapeCollection = MLNShapeCollection(shapes: annotations)
            source.shape = shapeCollection
        }
    }
    
    /**
     * Updates the preview line source with current polylines.
     */
    private func updatePreviewLineSource() {
        guard let style = mapView.style,
              let source = style.source(withIdentifier: previewLineSourceId) as? MLNShapeSource else {
            return
        }
        
        if previewLines.isEmpty {
            source.shape = nil
        } else {
            let polylines = Array(previewLines.values)
            let shapeCollection = MLNShapeCollection(shapes: polylines)
            source.shape = shapeCollection
        }
    }
    
    /**
     * Animates the appearance of a break point with a smooth transition.
     *
     * @param lineId The ID of the line the break point belongs to
     * @param coordinate The coordinate where the break point should appear
     */
    func animateBreakPointAppearance(lineId: String, coordinate: CLLocationCoordinate2D) {
        showBreakPoint(lineId: lineId, coordinate: coordinate)
        
        // Add a subtle animation by temporarily scaling the break point
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) { [weak self] in
            self?.animateBreakPointScale(lineId: lineId, scale: 1.2)
            
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.2) { [weak self] in
                self?.animateBreakPointScale(lineId: lineId, scale: 1.0)
            }
        }
    }
    
    /**
     * Animates the scale of a break point.
     *
     * @param lineId The ID of the line the break point belongs to
     * @param scale The scale factor to animate to
     */
    private func animateBreakPointScale(lineId: String, scale: Double) {
        guard let style = mapView.style,
              let layer = style.layer(withIdentifier: breakPointLayerId) as? MLNCircleStyleLayer else {
            return
        }
        
        let baseRadius = editingStyle["breakPointRadius"] as? Double ?? 8.0
        let animatedRadius = baseRadius * scale
        
        layer.circleRadius = NSExpression(forConstantValue: animatedRadius)
    }
    
    /**
     * Clears all visual editing elements from the map.
     */
    func clearAllVisualElements() {
        breakPointAnnotations.removeAll()
        previewLines.removeAll()
        updateBreakPointSource()
        updatePreviewLineSource()
        NSLog("\(EditablePolylineRenderer.TAG): Cleared all visual editing elements")
    }
    
    /**
     * Clears visual elements for a specific line.
     *
     * @param lineId The ID of the line to clear visual elements for
     */
    func clearVisualElementsForLine(lineId: String) {
        hideBreakPoint(lineId: lineId)
        hidePreviewLine(lineId: lineId)
        NSLog("\(EditablePolylineRenderer.TAG): Cleared visual elements for line \(lineId)")
    }
    
    /**
     * Parses a color string into a UIColor.
     *
     * @param colorString The color string to parse (e.g., "#FF0000")
     * @return The parsed UIColor, or red as default
     */
    private func parseColor(_ colorString: String) -> UIColor {
        var colorString = colorString.trimmingCharacters(in: .whitespacesAndNewlines)
        
        if colorString.hasPrefix("#") {
            colorString.removeFirst()
        }
        
        guard colorString.count == 6 else {
            NSLog("\(EditablePolylineRenderer.TAG): Invalid color string: \(colorString), using default red")
            return UIColor.red
        }
        
        var rgbValue: UInt64 = 0
        Scanner(string: colorString).scanHexInt64(&rgbValue)
        
        return UIColor(
            red: CGFloat((rgbValue & 0xFF0000) >> 16) / 255.0,
            green: CGFloat((rgbValue & 0x00FF00) >> 8) / 255.0,
            blue: CGFloat(rgbValue & 0x0000FF) / 255.0,
            alpha: 1.0
        )
    }
    
    /**
     * Gets statistics about the current visual state.
     *
     * @return A dictionary containing visual state statistics
     */
    func getVisualStatistics() -> [String: Any] {
        return [
            "activeBreakPoints": breakPointAnnotations.count,
            "activePreviewLines": previewLines.count,
            "currentStyle": editingStyle
        ]
    }
    
    /**
     * Cleans up resources when the renderer is no longer needed.
     */
    func cleanup() {
        clearAllVisualElements()
        
        // Remove layers and sources
        if let style = mapView.style {
            if style.layer(withIdentifier: breakPointLayerId) != nil {
                style.removeLayer(withIdentifier: breakPointLayerId)
            }
            if style.layer(withIdentifier: previewLineLayerId) != nil {
                style.removeLayer(withIdentifier: previewLineLayerId)
            }
            if style.source(withIdentifier: breakPointSourceId) != nil {
                style.removeSource(withIdentifier: breakPointSourceId)
            }
            if style.source(withIdentifier: previewLineSourceId) != nil {
                style.removeSource(withIdentifier: previewLineSourceId)
            }
        }
        
        NSLog("\(EditablePolylineRenderer.TAG): Cleanup complete")
    }
}

File: maplibre_gl/android/src/main/java/org/maplibre/maplibregl/EditablePolylineRenderer.java
// Copyright 2018 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.maplibre.maplibregl;

import android.graphics.Color;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.maplibre.android.geometry.LatLng;
import org.maplibre.android.maps.MapLibreMap;
import org.maplibre.android.style.layers.CircleLayer;
import org.maplibre.android.style.layers.PropertyFactory;
import org.maplibre.android.style.sources.GeoJsonSource;
import org.maplibre.geojson.Feature;
import org.maplibre.geojson.FeatureCollection;
import org.maplibre.geojson.Point;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Renders visual feedback elements for polyline editing operations.
 * 
 * This class manages the visual representation of:
 * - Break point markers during editing
 * - Visual state management for active editing sessions
 */
public class EditablePolylineRenderer {
    private static final String TAG = "EditablePolylineRenderer";
    
    // Layer and source IDs
    private static final String BREAK_POINT_SOURCE_ID = "polyline-editing-break-points";
    private static final String BREAK_POINT_LAYER_ID = "polyline-editing-break-points-layer";
    
    // Default styling
    private static final String DEFAULT_BREAK_POINT_COLOR = "#FF0000";
    private static final float DEFAULT_BREAK_POINT_RADIUS = 8.0f;
    private static final String DEFAULT_BREAK_POINT_BORDER_COLOR = "#FFFFFF";
    private static final float DEFAULT_BREAK_POINT_BORDER_WIDTH = 2.0f;
    
    private final MapLibreMap mapLibreMap;
    private final Map<String, BreakPointVisualState> activeBreakPoints;
    
    // Current styling configuration
    private String breakPointColor = DEFAULT_BREAK_POINT_COLOR;
    private float breakPointRadius = DEFAULT_BREAK_POINT_RADIUS;
    private String breakPointBorderColor = DEFAULT_BREAK_POINT_BORDER_COLOR;
    private float breakPointBorderWidth = DEFAULT_BREAK_POINT_BORDER_WIDTH;
    
    private boolean isInitialized = false;
    
    /**
     * Represents the visual state of a break point.
     */
    public static class BreakPointVisualState {
        public final String lineId;
        public final int pointIndex;
        public final LatLng location;
        public final boolean isVisible;
        public final boolean isDragging;
        public final long createdTime;
        
        public BreakPointVisualState(@NonNull String lineId, @NonNull LatLng location, 
                                   boolean isVisible, boolean isDragging) {
            this(lineId, -1, location, isVisible, isDragging);
        }

        public BreakPointVisualState(@NonNull String lineId, int pointIndex,
                                   @NonNull LatLng location,
                                   boolean isVisible, boolean isDragging) {
            this.lineId = lineId;
            this.pointIndex = pointIndex;
            this.location = location;
            this.isVisible = isVisible;
            this.isDragging = isDragging;
            this.createdTime = System.currentTimeMillis();
        }
        
        public BreakPointVisualState withLocation(@NonNull LatLng newLocation) {
            return new BreakPointVisualState(
                lineId, pointIndex, newLocation, isVisible, isDragging);
        }
        
        public BreakPointVisualState withDragging(boolean dragging) {
            return new BreakPointVisualState(
                lineId, pointIndex, location, isVisible, dragging);
        }
        
        public BreakPointVisualState withVisibility(boolean visible) {
            return new BreakPointVisualState(
                lineId, pointIndex, location, visible, isDragging);
        }
    }
    
    /**
     * Creates a new EditablePolylineRenderer.
     * 
     * @param mapLibreMap The MapLibre map instance
     */
    public EditablePolylineRenderer(@NonNull MapLibreMap mapLibreMap) {
        this.mapLibreMap = mapLibreMap;
        this.activeBreakPoints = new HashMap<>();
        
        Log.d(TAG, "EditablePolylineRenderer created");
    }
    
    /**
     * Initializes the renderer by adding necessary sources and layers to the map.
     * This should be called after the map style is loaded.
     * 
     * @return true if initialization was successful, false otherwise
     */
    public boolean initialize() {
        if (isInitialized) {
            Log.d(TAG, "Renderer already initialized");
            return true;
        }
        
        try {
            // Add break point source and layer
            if (!addBreakPointLayer()) {
                Log.e(TAG, "Failed to add break point layer");
                return false;
            }
            
            isInitialized = true;
            Log.d(TAG, "Renderer initialized successfully");
            return true;
            
        } catch (Exception e) {
            Log.e(TAG, "Error initializing renderer: " + e.getMessage(), e);
            return false;
        }
    }
    
    /**
     * Updates the visual styling configuration.
     * 
     * @param style Map containing style properties
     */
    public void updateStyle(@NonNull Map<String, Object> style) {
        Log.d(TAG, "Updating visual style: " + style);
        
        // Update break point styling
        if (style.containsKey("breakPointColor")) {
            breakPointColor = (String) style.get("breakPointColor");
        }
        if (style.containsKey("breakPointRadius")) {
            Object radius = style.get("breakPointRadius");
            if (radius instanceof Number) {
                breakPointRadius = ((Number) radius).floatValue();
            }
        }
        if (style.containsKey("breakPointBorderColor")) {
            breakPointBorderColor = (String) style.get("breakPointBorderColor");
        }
        if (style.containsKey("breakPointBorderWidth")) {
            Object width = style.get("breakPointBorderWidth");
            if (width instanceof Number) {
                breakPointBorderWidth = ((Number) width).floatValue();
            }
        }
        
        // Update break point layer styling
        if (isInitialized) {
            updateBreakPointLayerStyling();
        }
    }
    
    /**
     * Shows a break point marker at the specified location.
     * 
     * @param lineId The ID of the polyline
     * @param location The location of the break point
     */
    public void showBreakPoint(@NonNull String lineId, @NonNull LatLng location) {
        showBreakPoint(lineId, -1, location);
    }

    public void showBreakPoint(
            @NonNull String lineId, int pointIndex, @NonNull LatLng location) {
        if (!isInitialized) {
            Log.w(TAG, "Renderer not initialized, cannot show break point");
            return;
        }
        
        Log.d(TAG, "Showing break point for line " + lineId + " at " + location);
        
        try {
            BreakPointVisualState state = new BreakPointVisualState(
                lineId, pointIndex, location, true, false);
            activeBreakPoints.put(breakPointKey(lineId, pointIndex), state);
            updateBreakPointSource();
            
        } catch (Exception e) {
            Log.e(TAG, "Error showing break point: " + e.getMessage(), e);
        }
    }
    
    /**
     * Updates the location of an existing break point marker.
     * 
     * @param lineId The ID of the polyline
     * @param newLocation The new location of the break point
     */
    public void updateBreakPoint(@NonNull String lineId, @NonNull LatLng newLocation) {
        updateBreakPoint(lineId, -1, newLocation);
    }

    public void updateBreakPoint(
            @NonNull String lineId, int pointIndex, @NonNull LatLng newLocation) {
        if (!isInitialized) {
            Log.w(TAG, "Renderer not initialized, cannot update break point");
            return;
        }
        
        BreakPointVisualState currentState =
            activeBreakPoints.get(breakPointKey(lineId, pointIndex));
        if (currentState == null) {
            Log.w(TAG, "No active break point found for line " + lineId);
            return;
        }
        
        try {
            BreakPointVisualState updatedState = currentState.withLocation(newLocation);
            activeBreakPoints.put(breakPointKey(lineId, pointIndex), updatedState);
            updateBreakPointSource();
            
            Log.d(TAG, "Updated break point for line " + lineId + " to " + newLocation);
            
        } catch (Exception e) {
            Log.e(TAG, "Error updating break point: " + e.getMessage(), e);
        }
    }
    
    /**
     * Sets the dragging state of a break point marker.
     * 
     * @param lineId The ID of the polyline
     * @param isDragging Whether the break point is being dragged
     */
    public void setBreakPointDragging(@NonNull String lineId, boolean isDragging) {
        setBreakPointDragging(lineId, -1, isDragging);
    }

    public void setBreakPointDragging(
            @NonNull String lineId, int pointIndex, boolean isDragging) {
        if (!isInitialized) {
            return;
        }
        
        BreakPointVisualState currentState =
            activeBreakPoints.get(breakPointKey(lineId, pointIndex));
        if (currentState == null) {
            return;
        }
        
        try {
            BreakPointVisualState updatedState = currentState.withDragging(isDragging);
            activeBreakPoints.put(breakPointKey(lineId, pointIndex), updatedState);
            updateBreakPointSource();
            
        } catch (Exception e) {
            Log.e(TAG, "Error setting break point dragging state: " + e.getMessage(), e);
        }
    }
    
    /**
     * Hides the break point marker for the specified polyline.
     * 
     * @param lineId The ID of the polyline
     */
    public void hideBreakPoint(@NonNull String lineId) {
        if (!isInitialized) {
            return;
        }
        
        try {
            activeBreakPoints.entrySet().removeIf(
                entry -> entry.getValue().lineId.equals(lineId));
            updateBreakPointSource();
            
            Log.d(TAG, "Hidden break point for line " + lineId);
            
        } catch (Exception e) {
            Log.e(TAG, "Error hiding break point: " + e.getMessage(), e);
        }
    }

    public void syncBreakPoints(
            @NonNull String lineId,
            @NonNull List<LatLng> coordinates,
            @NonNull java.util.Set<Integer> lockedIndices) {
        if (!isInitialized) return;
        activeBreakPoints.entrySet().removeIf(
            entry -> entry.getValue().lineId.equals(lineId));
        for (int index = 1; index < coordinates.size() - 1; index++) {
            if (lockedIndices.contains(index)) continue;
            BreakPointVisualState state = new BreakPointVisualState(
                lineId, index, coordinates.get(index), true, false);
            activeBreakPoints.put(breakPointKey(lineId, index), state);
        }
        updateBreakPointSource();
    }

    private String breakPointKey(@NonNull String lineId, int pointIndex) {
        return lineId + ":" + pointIndex;
    }
    
    /**
     * Clears all visual feedback elements.
     */
    public void clearAll() {
        if (!isInitialized) {
            return;
        }
        
        try {
            int breakPointCount = activeBreakPoints.size();
            
            activeBreakPoints.clear();
            updateBreakPointSource();
            
            Log.d(TAG, "Cleared " + breakPointCount + " break points");
            
        } catch (Exception e) {
            Log.e(TAG, "Error clearing visual feedback: " + e.getMessage(), e);
        }
    }
    
    /**
     * Gets the number of active break points.
     * 
     * @return The number of active break points
     */
    public int getActiveBreakPointCount() {
        return activeBreakPoints.size();
    }
    
    /**
     * Checks if the renderer is initialized.
     * 
     * @return true if initialized, false otherwise
     */
    public boolean isInitialized() {
        return isInitialized;
    }
    
    /**
     * Adds the break point source and layer to the map.
     */
    private boolean addBreakPointLayer() {
        try {
            // Add empty source
            GeoJsonSource breakPointSource = new GeoJsonSource(BREAK_POINT_SOURCE_ID, 
                FeatureCollection.fromFeatures(new ArrayList<>()));
            mapLibreMap.getStyle().addSource(breakPointSource);
            
            // Add circle layer for break points
            CircleLayer breakPointLayer = new CircleLayer(BREAK_POINT_LAYER_ID, BREAK_POINT_SOURCE_ID);
            breakPointLayer.setProperties(
                PropertyFactory.circleRadius(breakPointRadius),
                PropertyFactory.circleColor(Color.parseColor(breakPointColor)),
                PropertyFactory.circleStrokeColor(Color.parseColor(breakPointBorderColor)),
                PropertyFactory.circleStrokeWidth(breakPointBorderWidth)
            );
            
            mapLibreMap.getStyle().addLayer(breakPointLayer);
            
            Log.d(TAG, "Added break point layer");
            return true;
            
        } catch (Exception e) {
            Log.e(TAG, "Error adding break point layer: " + e.getMessage(), e);
            return false;
        }
    }
    
    /**
     * Updates the styling of break point layer.
     */
    private void updateBreakPointLayerStyling() {
        try {
            // Update break point layer styling
            CircleLayer breakPointLayer = (CircleLayer) mapLibreMap.getStyle().getLayer(BREAK_POINT_LAYER_ID);
            if (breakPointLayer != null) {
                breakPointLayer.setProperties(
                    PropertyFactory.circleRadius(breakPointRadius),
                    PropertyFactory.circleColor(Color.parseColor(breakPointColor)),
                    PropertyFactory.circleStrokeColor(Color.parseColor(breakPointBorderColor)),
                    PropertyFactory.circleStrokeWidth(breakPointBorderWidth)
                );
            }
            
            Log.d(TAG, "Updated break point layer styling");
            
        } catch (Exception e) {
            Log.e(TAG, "Error updating layer styling: " + e.getMessage(), e);
        }
    }
    
    /**
     * Updates the break point source with current break point data.
     */
    private void updateBreakPointSource() {
        try {
            List<Feature> features = new ArrayList<>();
            
            for (BreakPointVisualState state : activeBreakPoints.values()) {
                if (state.isVisible) {
                    Point point = Point.fromLngLat(state.location.getLongitude(), state.location.getLatitude());
                    Feature feature = Feature.fromGeometry(point);
                    feature.addStringProperty("lineId", state.lineId);
                    feature.addNumberProperty("pointIndex", state.pointIndex);
                    feature.addBooleanProperty("isDragging", state.isDragging);
                    features.add(feature);
                }
            }
            
            GeoJsonSource source = (GeoJsonSource) mapLibreMap.getStyle().getSource(BREAK_POINT_SOURCE_ID);
            if (source != null) {
                source.setGeoJson(FeatureCollection.fromFeatures(features));
            }
            
        } catch (Exception e) {
            Log.e(TAG, "Error updating break point source: " + e.getMessage(), e);
        }
    }
    
    /**
     * Logs the current state of the renderer for debugging.
     */
    public void logState() {
        Log.d(TAG, "EditablePolylineRenderer state:");
        Log.d(TAG, "  Initialized: " + isInitialized);
        Log.d(TAG, "  Active break points: " + activeBreakPoints.size());
        
        for (Map.Entry<String, BreakPointVisualState> entry : activeBreakPoints.entrySet()) {
            BreakPointVisualState state = entry.getValue();
            Log.d(TAG, "    Break point " + entry.getKey() + ": visible=" + state.isVisible + 
                      ", dragging=" + state.isDragging + ", location=" + state.location);
        }
    }
}
